tests/test_run_reference_vmm.py

```diff
@@ -93,33 +95,59 @@ def shutdown(vmm_process):
     # TimeoutExpired exception and fail the test.
     vmm_process.wait(timeout=3)
 
+def setup_stdout_nonblocking(vmm_process):
+
+    ## We'll need to do non-blocking I/O with the underlying sub-process since
+    # we cannot use `communicate`, because `communicate` would close the
+    # `stdin` that we later want to use to `shutdown`, to do that by hand,
+    # we set `vmm_process.stdout` to non-blocking
+    # Then we can use `os.read` that would raise `BlockingIOError`
+
+    # FIXME: This should NOT be required to be done on every call, do it when we
+    #        'Class'ify the test case
+    flags = fcntl.fcntl(vmm_process.stdout, fcntl.F_GETFL)
+    fcntl.fcntl(vmm_process.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+def expect_string(vmm_process, expected_string, timeout=TEST_TIMEOUT):
+
+    setup_stdout_nonblocking(vmm_process)
+
+    # No. of seconds after which we'll give up
+    giveup_after = timeout
+    then = time.time()
+    found = False
+    # This is required because the pattern we are expecting might get split across two reads
+    all_data = bytes()
+    while not found:
+        try:
+            x = os.read(vmm_process.stdout.fileno(), 4096)
```

Comment:
Is it possible to still use `readline`? It could make the code more readable than manually splitting by `\r\n`.

Reply:
Unfortunately no. I tried to use `stdout.readline` but it does not throw the `BlockingIOError` exception as above and simply blocks when it can't find a new-line. 

I agree this whole part is somewhat difficult to understand (if not very ugly), but unfortunately, I cannot think of a cleaner way of doing this. This happens because `Popen.communicate` closes the `stdin` that we want to use later, so effectively it means we have to do all the communication with the process by hand and hence all the trickery is required!

Reply:
We can leave it as is, no problem.